lib/cmsis_rtos_v2/thread.c

```diff
@@ -83,36 +100,69 @@ osThreadId_t osThreadNew(osThreadFunc_t threadfunc, void *arg,
 	s32_t prio;
 	struct cv2_thread *tid;
 	static u32_t one_time;
+	size_t stack_size;
+	u32_t this_thread_num;
+	u32_t this_thread_num_dynamic;
+
+	if (k_is_in_isr()) {
+		return NULL;
+	}
+
+	if (thread_num >= CONFIG_CMSIS_V2_THREAD_MAX_COUNT) {
+		__ASSERT(thread_num < CONFIG_CMSIS_V2_THREAD_MAX_COUNT,
+			 "exceeded maximum thread count");
+		return NULL;
+	}
+
+	if (attr == NULL) {
+		attr = &init_thread_attrs;
+	}
+
+	if ((attr->stack_mem == NULL) && (thread_num_dynamic >=
+					  CONFIG_CMSIS_V2_THREAD_DYNAMIC_MAX_COUNT)) {
+		__ASSERT(thread_num < CONFIG_CMSIS_V2_THREAD_MAX_COUNT,
+			 "exceeded maximum dynamic thread count");
+		return NULL;
+	}
 
 	BUILD_ASSERT_MSG(osPriorityISR <= CONFIG_NUM_PREEMPT_PRIORITIES,
 			 "Configure NUM_PREEMPT_PRIORITIES to at least osPriorityISR");
 
-	__ASSERT(attr->stack_size <= CONFIG_CMSIS_V2_THREAD_MAX_STACK_SIZE,
-		 "invalid stack size\n");
+	BUILD_ASSERT_MSG(CONFIG_CMSIS_V2_THREAD_DYNAMIC_MAX_COUNT <=
+			 CONFIG_CMSIS_V2_THREAD_MAX_COUNT,
+			 "Number of dynamic threads cannot exceed max number of threads.");
 
-	__ASSERT(thread_num <= CONFIG_CMSIS_V2_THREAD_MAX_COUNT,
-		 "Exceeded max number of threads\n");
+	BUILD_ASSERT_MSG(CONFIG_CMSIS_V2_THREAD_DYNAMIC_STACK_SIZE <=
+			 CONFIG_CMSIS_V2_THREAD_MAX_STACK_SIZE,
+			 "Default dynamic thread stack size cannot exceed max stack size");
 
-	__ASSERT(attr != NULL,
-		 "Zephyr requirement: Pass attributes explicitly for ThreadNew\n");
+	__ASSERT((attr->stack_mem == NULL) || (attr->stack_size > 0),
+		 "stack_size must be greater than 0 if stack_mem is not NULL\n");
```

Comment:
The original `__ASSERT` is as intended.

It is stating that if `attr->stack_mem` is `not NULL` then `attr->stack_size` must be greater than 0.

The truth table would be:

```
stack_mem    stack_size    result?

  NULL           0          true
  NULL           >0         true
 Not NULL        0          false
 Not NULL        >0         true
````

From the above truth table the following boolean expression can be derived:

`state_mem == NULL` **OR** `stack_size > 0`

The only other way to do this would be to break it up into nested if-statements.

Opinion?

Reply:
Checking for attr->stack_size when attr->stack_mem is NULL doesn't make any sense as we don't care about it.  So, I don't think you should consider the first 2 rows in your truth table. This then boils down to the condition I specified. Moreover, attr->stack_mem is being checked for NULL in "if else" at line 162 and 168 and is hence covered.

Reply:
> However the condition you specified wouldn't hold true when `stack_mem` is `NULL` thus causing the `ASSERT` to be `false` and halting execution even though the attributes are valid. Logical **OR** is a lazy operator so only the first operand is evaluated if it is `true`, meaning the `stack_size` is then never checked if `stack_mem` is `NULL`.
> 
> While `stack_mem` is checked for being `NULL` later on in the function this is only to ascertain whether dynamic allocation should be attempted or not, it doesn't protect again nor detect the situation where a valid `stack_mem` pointer has been passed in but the `stack_size` is invalid i.e. 0.
> 
> Maybe a clearer implementation would be:
> 
> ```
> if (attr->stack_mem != NULL) {
>     if (attr->stack_size == 0) {
>         return NULL; /* stack_size must be non-zero if stack_mem is set. */
>     }
> }
I agree. Please go ahead with this implementation
> ```
> This more explicitly avoids any check of `stack_size` if `stack_mem` is `NULL` and still causes the function to fail (and without an `ASSERT` as suggested in another comment).